Credit card data

In [ ]:
import pandas as pd
data = pd.read_csv('/content/drive/MyDrive/PA/creditcard.csv')
data.dropna(inplace=True)

In [ ]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

In [ ]:
display(data.head())

,Time,V1,V2,V3,V4,V5,V6,V7,V8,V9,...,V21,V22,V23,V24,V25,V26,V27,V28,Amount,Class
0,0.0,-1.359807,-0.072781,2.536347,1.378155,-0.338321,0.462388,0.239599,0.098698,0.363787,...,-0.018307,0.277838,-0.110474,0.066928,0.128539,-0.189115,0.133558,-0.021053,149.62,0.0
1,0.0,1.191857,0.266151,0.166480,0.448154,0.060018,-0.082361,-0.078803,0.085102,-0.255425,...,-0.225775,-0.638672,0.101288,-0.339846,0.167170,0.125895,-0.008983,0.014724,2.69,0.0
2,1.0,-1.358354,-1.340163,1.773209,0.379780,-0.503198,1.800499,0.791461,0.247676,-1.514654,...,0.247998,0.771679,0.909412,-0.689281,-0.327642,-0.139097,-0.055353,-0.059752,378.66,0.0
3,1.0,-0.966272,-0.185226,1.792993,-0.863291,-0.010309,1.247203,0.237609,0.377436,-1.387024,...,-0.108300,0.005274,-0.190321,-1.175575,0.647376,-0.221929,0.062723,0.061458,123.50,0.0
4,2.0,-1.158233,0.877737,1.548718,0.403034,-0.407193,0.095921,0.592941,-0.270533,0.817739,...,-0.009431,0.798278,-0.137458,0.141267,-0.206010,0.502292,0.219422,0.215153,69.99,0.0


In [ ]:
print(data.info)

<bound method DataFrame.info of         Time        V1        V2        V3        V4        V5        V6  \
0        0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388   
1        0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361   
2        1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499   
3        1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203   
4        2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921   
...      ...       ...       ...       ...       ...       ...       ...   
1439  1114.0 -1.132962  0.204704  1.588581  0.079664  0.388785  0.118162   
1440  1117.0  0.956173 -0.403827  0.348913  0.625389 -0.672878 -0.714691   
1441  1117.0  1.240985  0.096546  0.369956  0.440568 -0.381717 -0.513028   
1442  1117.0 -1.380039 -0.130959  0.125171 -2.168485  0.548425 -1.724708   
1443  1120.0  1.380183 -1.162585  0.549074 -1.559286 -1.303020  0.088948   

            V7        V8        V9  ...       V21      

In [ ]:
x = data.drop('Class', axis=1)
y = data['Class']

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=42, stratify=y)

In [ ]:
import statsmodels.formula.api as smf
import pandas as pd
import io


In [ ]:
mtcars_data = """model,mpg,cyl,disp,hp,drat,wt,qsec,vs,am,gear,carb
Mazda RX4,21,6,160,110,3.90,2.620,16.46,0,1,4,4
Datsun 710,22.8,4,199,93,3.85,2.32,18.61,1,1,4,1
Hornet 4 Drive,21.4,6,258,110,3.08,3.215,19.44,1,0,3,3
Valiant,19.8,6,22105,2.76,3.46,46,20.22,1,8,3,1
Merc 2480,24.4,4,146.7,62,3.69,3.19,20,1,0,4,2"""

In [ ]:
df = pd.read_csv(io.StringIO(mtcars_data),sep=',')

In [ ]:
results = smf.ols('mpg ~ disp + hp + cyl', df).fit()

In [ ]:
results.summary()

/usr/local/lib/python3.12/dist-packages/statsmodels/stats/stattools.py:74: ValueWarning: omni_normtest is not valid with less than 8 observations; 5 samples were given.
  warn("omni_normtest is not valid with less than 8 observations; %i "


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    mpg   R-squared:                       0.993
Model:                            OLS   Adj. R-squared:                  0.970
Method:                 Least Squares   F-statistic:                     44.68
Date:                Thu, 18 Sep 2025   Prob (F-statistic):              0.109
Time:                        03:58:02   Log-Likelihood:                 2.8731
No. Observations:                   5   AIC:                             2.254
Df Residuals:                       1   BIC:                            0.6915
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     29.0683      0.796     36.497      0.017      18.948      39.188
disp          -0.0003   6.96e-05     -4.500      0.139      -0.001       0.001
hp            -0.0509      0.014     -3.692      0.168      -0.226       0.124
cyl           -0.3679      0.272     -1.354      0.405      -3.821       3.085
==============================================================================
Omnibus:                          nan   Durbin-Watson:                   1.507
Prob(Omnibus):                    nan   Jarque-Bera (JB):                0.052
Skew:                           0.011   Prob(JB):                        0.974
Kurtosis:                       2.499   Cond. No.                     5.83e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 5.83e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

### Statistical Interpretation of OLS Regression Results

1.  **R-squared (0.993) and Adjusted R-squared (0.970):**
    *   **Interpretation:** These values are extremely high, suggesting that approximately 99.3% of the variance in `mpg` (miles per gallon) can be explained by the independent variables `disp` (engine displacement), `hp` (horsepower), and `cyl` (number of cylinders). The adjusted R-squared accounts for the number of predictors, and it's also very high.
    *   **Caution:** With only 5 observations (as indicated by "No. Observations: 5"), such a high R-squared is highly suspicious and often indicates **overfitting**. The model is likely explaining the noise in the very small dataset rather than true underlying relationships.

2.  **F-statistic (44.68) and Prob (F-statistic) (0.109):**
    *   **Interpretation:** The F-statistic tests the overall significance of the regression model. The p-value of 0.109 is greater than the common significance level of 0.05. This means that, **overall, the model is not statistically significant at the 5% level.** This contradicts the very high R-squared and further suggests that the model's predictive power is not reliable with this limited data.

3.  **Coefficients (coef) and P>|t| values:**
    *   **Intercept (29.0683, P=0.017):** This is statistically significant (P < 0.05). It suggests that when `disp`, `hp`, and `cyl` are all zero, `mpg` is predicted to be 29.0683. However, zero values for these car characteristics are not realistic, so the intercept's practical interpretation is limited.
    *   **disp (-0.0003, P=0.139):** Not statistically significant (P > 0.05). The coefficient suggests a very small negative relationship: for every one-unit increase in `disp`, `mpg` decreases by 0.0003 units, *holding `hp` and `cyl` constant*. Due to non-significance, we cannot confidently conclude this is a real effect.
    *   **hp (-0.0509, P=0.168):** Not statistically significant (P > 0.05). The coefficient suggests that for every one-unit increase in `hp`, `mpg` decreases by 0.0509 units, *holding `disp` and `cyl` constant*. Again, due to non-significance, this is not a reliable finding.
    *   **cyl (-0.3679, P=0.405):** Not statistically significant (P > 0.05). The coefficient suggests that for every one-unit increase in `cyl`, `mpg` decreases by 0.3679 units, *holding `disp` and `hp` constant*. This is also not a reliable finding.

4.  **Notes - Condition Number (5.83e+04):**
    *   **Interpretation:** A very large condition number indicates **strong multicollinearity** among your independent variables (`disp`, `hp`, `cyl`). This means these variables are highly correlated with each other, which can make the individual coefficient estimates unstable, unreliable, and difficult to interpret independently.

**Overall Statistical Conclusion:** The model suffers from severe limitations due to an extremely small sample size (only 5 observations) and strong multicollinearity. While the R-squared is high, the overall model is not statistically significant, and none of the individual predictors are statistically significant. Therefore, **the statistical findings of this model are not reliable or robust.**

### Business Interpretation of OLS Regression Results (with extreme caution)

Given the unreliability of the statistical results, any business interpretations derived from this specific model should be treated with extreme skepticism and **not used for decision-making.** However, *if* the model were robust and statistically significant, here's how one *might* interpret the findings from a business perspective:

*   **General Relationship:** The model attempts to quantify how engine characteristics (`disp`, `hp`, `cyl`) impact fuel efficiency (`mpg`). In the automotive industry, understanding these relationships is crucial for design, marketing, and competitive positioning.

*   **Impact of Engine Size (Displacement, `disp`):** If the negative coefficient for `disp` were statistically significant, it would suggest that larger engine displacement generally leads to lower fuel economy. For car manufacturers, this would reinforce the trend towards smaller, more efficient engines (e.g., downsizing with turbocharging) to meet fuel efficiency standards and consumer demand for economical vehicles.

*   **Impact of Power (Horsepower, `hp`):** Similarly, if the negative coefficient for `hp` were significant, it would imply a trade-off: more powerful engines consume more fuel. This is a common design challenge, balancing performance with fuel economy. Businesses might use this to target different market segments (e.g., performance cars vs. economy cars).

*   **Impact of Cylinders (`cyl`):** If the negative coefficient for `cyl` were significant, it would suggest that cars with more cylinders tend to be less fuel-efficient. This is often linked to larger engine sizes and greater internal friction, reinforcing the move towards 4-cylinder or 3-cylinder engines in many mainstream vehicles.

**Business Actions/Insights (Based on reliable data, not these results):**

*   **Product Development:** Engineers and product planners would use such insights to optimize engine design for desired `mpg` targets, balancing performance, cost, and regulatory requirements.
*   **Marketing & Sales:** Marketing teams could highlight superior fuel efficiency (if achieved) as a key selling point or position powerful vehicles for specific buyer segments who prioritize performance over `mpg`.
*   **Strategic Decisions:** Long-term business strategy would consider these relationships for investment in R&D (e.g., hybrid, electric, or advanced internal combustion technologies).

**Primary Business Takeaway from *This Specific Analysis*:**

**The most important business conclusion from this analysis is that the current dataset is insufficient to draw reliable conclusions.** Before making any business decisions, a much larger dataset is required to build a statistically sound and robust model. Addressing multicollinearity would also be necessary to isolate the individual effects of `disp`, `hp`, and `cyl` more accurately.

In [ ]:
print(results.wald_test('hp = 0, cyl = 0'))

<F test: F=array([[37.73778711]]), p=0.11435064713135343, df_denom=1, df_num=2>


/usr/local/lib/python3.12/dist-packages/statsmodels/base/model.py:1912: FutureWarning: The behavior of wald_test will change after 0.14 to returning scalar test statistic values. To get the future behavior now, set scalar to True. To silence this message while retaining the legacy behavior, set scalar to False.
  warnings.warn(
